File: src/main/java/com/github/unldenis/hologram/TextLine.java
/*
 * Hologram-Lib - Asynchronous, high-performance Minecraft Hologram
 * library for 1.8-1.18 servers.
 * Copyright (C) unldenis <https://github.com/unldenis>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.unldenis.hologram;

import com.github.unldenis.hologram.packet.PacketsFactory;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

public class TextLine extends AbstractLine<String> {

  private boolean isEmpty;

  public TextLine(@NotNull Hologram hologram, @NotNull String obj) {
    super(hologram, obj);
  }

  @Override
  protected void show(@NotNull Player player) {
    isEmpty = obj.isEmpty();
    if (!isEmpty) {
      super.show(player);
      PacketsFactory.get()
          .metadataPacket(entityID, obj, player, hologram.getPlaceholders(), true, true)
          .send(player);
    }
  }

  @Override
  protected void update(@NotNull Player player) {
    byte spawnBefore = (byte) ((isEmpty ? 1 : 0) | (obj.isEmpty() ? 1 : 0) << 1);
    /*
      0x00  = is already showed
      0x01  = is hided but now has changed
      0x02  = is already showed but is empty
      0x03  = is hided and isn't changed
     */
    switch (spawnBefore) {
      case 0x03:
        break;
      case 0x02:
        super.hide(player);
        isEmpty = true;
        break;
      case 0x01:
        super.show(player);
        isEmpty = false;
      case 0x00:
        PacketsFactory.get()
            .metadataPacket(entityID, obj, player, hologram.getPlaceholders(), spawnBefore == 0x01,
                spawnBefore == 0x01)
            .send(player);
    }


  }

  @Override
  @NotNull
  public String get() {
    return obj;
  }

  /**
   * Parse the possible placeholder for the player.
   *
   * @param player - the target player
   * @return the parsed strings
   */
  @NotNull
  public String parse(Player player) {
    return hologram.getPlaceholders().parse(obj, player);
  }

}

File: src/main/java/org/holoeasy/line/TextLine.kt
package org.holoeasy.line

import org.holoeasy.builder.interfaces.PlayerFun
import org.bukkit.Location
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import org.holoeasy.ext.send
import org.holoeasy.packet.IPacket
import org.holoeasy.util.AABB

class TextLine(plugin: Plugin, obj: String, override val args: Array<PlayerFun>? = null,
               override val clickable: Boolean = false
) : ITextLine {

    private val line: Line = Line(plugin, EntityType.ARMOR_STAND)

    override var obj: String = ""

    var clickEvent : ClickEvent? = null

    init {

        if (args == null) {
            this.obj = obj
        } else {
            this.obj = obj.replace("{}", "%s")
        }
    }

    var hitbox: AABB? = null
        private set
    private var isEmpty = false

    override val textLine: TextLine
        get() = this

    override fun parse(player: Player): String {
        if (args == null) {
            return obj
        }
        val res = arrayOfNulls<Any>(args.size)
        for (i in args.indices) {
            res[i] = args[i](player)
        }

        return String.format(obj, args = res)
    }

    override fun onClick(clickEvent: ClickEvent) {
        this.clickEvent = clickEvent
    }

    override val plugin: Plugin
        get() = line.plugin
    override val type: ILine.Type
        get() = ILine.Type.TEXT_LINE
    override val entityId: Int
        get() = line.entityID
    override val location: Location?
        get() = line.location

    override fun setLocation(value: Location) {
        line.location = value
        if (clickable) {
            val chars = obj.length.toDouble()
            val size = 0.105
            val dist = size * (chars / 2.0)

            hitbox = AABB(
                AABB.Vec3D(-dist, -0.039, -dist),
                AABB.Vec3D(dist, +0.039, dist)
            ).also {
                it.translate(AABB.Vec3D.fromLocation(value.clone().add(0.0, 1.40, 0.0)))
            }
        }
    }

    override fun hide(player: Player) {
        line.destroy(player)
    }

    override fun teleport(player: Player) {
        line.teleport(player)
    }

    override fun show(player: Player) {
        isEmpty = obj.isEmpty()
        if (!isEmpty) {
            line.spawn(player)
            val packet = IPacket.get(IPacket.Type.METADATA_TEXT)
                .metadata(entityId, parse(player))
            packet.send(player)
        }
    }

    override fun update(player: Player) {
        val spawnBefore = ((if (isEmpty) 1 else 0) or ((if (obj.isEmpty()) 1 else 0) shl 1))
        /*  0x00  = is already showed
            0x01  = is hided but now has changed
            0x02  = is already showed but is empty
            0x03  = is hided and isn't changed      */
        when (spawnBefore) {
            0x03 -> {}
            0x02 -> {
                line.destroy(player)
                isEmpty = true
            }

            0x01 -> {
                line.spawn(player)
                isEmpty = false
                IPacket.get(IPacket.Type.METADATA_TEXT)
                    .metadata(entityId, parse(player)).send(player)
            }

            0x00 ->
                IPacket.get(IPacket.Type.METADATA_TEXT)
                    .metadata(entityId, parse(player), invisible = false)
                    .send(player)
        }
    }


}